src/Core/Form/ChoiceProvider/ThemeByNameWithEmailsChoiceProvider.php

```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * 2007-2019 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Form\ChoiceProvider;
+
+use PrestaShop\PrestaShop\Core\Addon\Theme\Theme;
+use PrestaShop\PrestaShop\Core\Addon\Theme\ThemeCollection;
+use PrestaShop\PrestaShop\Core\Form\FormChoiceProviderInterface;
+
+/**
+ * Class ThemeByNameChoiceProvider provides theme choices with name values, but it
+ * filters themes which haven't overridden any email templates.
+ */
+final class ThemeByNameWithEmailsChoiceProvider implements FormChoiceProviderInterface
+{
+    /**
+     * @var ThemeCollection collection of themes
+     */
+    private $themeCollection;
+
+    /**
+     * @param ThemeCollection $themeCollection
+     */
+    public function __construct(ThemeCollection $themeCollection)
+    {
+        $this->themeCollection = $themeCollection;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getChoices()
+    {
+        $themeChoices = [];
+
+        /** @var Theme $theme */
+        foreach ($this->themeCollection as $theme) {
+            $coreMailsFolder = $theme->getDirectory() . '/mails';
+            $modulesMailFolder = $theme->getDirectory() . '/modules';
+            if (is_dir($coreMailsFolder) && is_dir($modulesMailFolder)) {
```

Comment:
Why `&&` we can have modules without mails directories :thinking:

Reply:
Yes but we don't check if the sub folder contain mails only if the theme folder contains the `modules` folder, partly because `MailTemplateGenerator` always checks it:
https://github.com/PrestaShop/PrestaShop/blob/06157a9c20628aa5e6549e5feaf704fe10548b11/src/Core/MailTemplate/MailTemplateGenerator.php#L89

Which require that both `coreMailsFolder` and `modulesMailFolder` exist. But then maybe we can change the behavior:
- listed themes are the the themes containing `mails` or `modules` folder
- when overwriting a theme we need to check if the folders exist and fallback on the default folder if they don't (in https://github.com/PrestaShop/PrestaShop/blob/06157a9c20628aa5e6549e5feaf704fe10548b11/src/PrestaShopBundle/Controller/Admin/Improve/Design/MailThemeController.php#L112 )

Reply:
@PierreRambaud all right for you ?

Reply:
Yes, I think it will be done in a new PR :wink: